### PR TITLE
Add processor cache management proposal

### PR DESCRIPTION
**Pull request by linericyang**
This is my first proposal to community repo, please advise if something need to change, such as process steps, format, etc. Thanks.

### CODE REVIEW

File: contributors/design-proposals/node/cache-management.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  launched containers except for a streaming application such as a video player
+  running first as  low priority and next as high priority. In the former
+  scenarios, cache disruption is limited to the bottom half while in the latter
+  it has full use of the cache and thus affects adversely more co-resident
+  container workloads. 
+
+## Implementation Roadmap
+
+### Phase 1
+V 1.10 or 1.11 to include the above experimental cache allocation.
```
Comment (ConnorDoyle) [inline contributors/design-proposals/node/cache-management.md]: v1.11 at the earliest

**Comment by ConnorDoyle** [comment]
/sig node
/kind feature
@kubernetes/sig-node-proposals

File: contributors/design-proposals/node/cache-management.md (hunk 1 above)
Comment (LinEricYang) [reply contributors/design-proposals/node/cache-management.md]: Thanks for pointing out this. Done.

**Comment by jeremyeder** [comment]
The design does not seem to mention integration with the existing system-reserved scaffolding.  Had you considered a basic use-case, where we segment the cache into partitions (perhaps proportional in size) based on the use of system-reserved?

**Comment by LinEricYang** [comment]
@jeremyeder Thanks for your comments. Actually segment the cache in size is the very first thought when we design the cache allocation strategy. However, cache allocation requires contiguous cache ways, it introduces fragmentation.  Current we don't have a perfect solution to resolve it without bringing in additional cache warm-up latency. So for "static" policy, we are seeking a simple solution which will meet most of requirements. And will try to improve it in "dynamic" policy by making it be able to track and understand workload requirements and do the proper action for it. Any thoughts?

**Comment by derekwaynecarr** [comment]
i am going to defer reviewing this too much until 1.11 re-opens.

File: contributors/design-proposals/node/cache-management.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+The original [cpu manager design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/cpu-manager.md) alluded to extending its scope
+to control processor cache allocation to obtain more predictable performance. 
+The CPU Manager intercepts all container related actions, in particular add,
+remove, start, and state update. Below we discuss changes for cache allocation. 
+
+### Pod QoS Class
+
+Kubernetes supports a notion of [pod QoS](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/), the higher the value, the higher
+the priority. We treat the Guaranteed QOS pod as deserving of more cache
```
Comment (thockin) [inline contributors/design-proposals/node/cache-management.md]: I'm concerned that QoS is the wrong indicator.  It seems completely likely that batch work has limits specified and it seems equally likely (maybe more) that latency-sensitive work might not specify CPU limit (or might set it higher than request).  This latter is the default operating mode for Borg, and I explicitly advise people to not limit CPU unless they have very careful benchmarks and deep understanding of the implications.  It does not work how one would naively expect.

I think we may want some other indicator for this, e.g. an "archetype" selector or similar.  The nice thing about considering this as a new thing is that we can quota it on its own.  Having a limit on # of cores of latency-sensitive work seems useful.

File: contributors/design-proposals/node/cache-management.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+### Alternative Cache Configuration
+
+Instead of just two (or more) overlapping COS, one might ask why not partition
+the cache and assign each exclusively to a CPU core, along the lines of CPU
+pinning. The underlying cache management library provides such support, but in
+this experimental  release, we opt for overlapping pools because of the greater
+sharing they provide (leading to better best-effort performance) yet providing
+less contention for workloads that have a higher QOS requirement. Further, when
+workloads are pinned to a CPU core, they already have exclusive access to the
+core’s L1 and L2 cache.
```
Comment (vishh) [inline contributors/design-proposals/node/cache-management.md]: How does hyperthreading impact this assumption?

File: contributors/design-proposals/node/cache-management.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  workload is assigned a priority equal to high and all other workloads are
+  assigned priority low with either “Static” or “Dynamic” policy.
+- Perceptible difference in performance, all things being equal with respect to
+  launched containers except for a streaming application such as a video player
+  running first as  low priority and next as high priority. In the former
+  scenarios, cache disruption is limited to the bottom half while in the latter
+  it has full use of the cache and thus affects adversely more co-resident
+  container workloads. 
+
+## Implementation Roadmap
```
Comment (vishh) [inline contributors/design-proposals/node/cache-management.md]: I'd like to see monitoring of cache allocations and memory bandwidth prior to adding a feature to restrict it. How can a k8s user know if they need to consume cache partitioning?

**Comment by vishh** [comment]
@LinEricYang I find manual cache partitioning between `shared` and `reserved` a bit tedious since a user will then have to carefully plan capacity and scheduling. Would it be possible to dynamically adjust CLOS sizes based on usage metrics and policies (max limits for batch for example)? 
AFAIK, users are yet to run batch workloads at scale alongside their web services (or databases) on k8s and so as you mention this feature will be isolated to specialized workloads. Can you post some metrics on the impact of CAT on your NFV workloads? I'm interested in exploring the impact of this feature on ML workloads. Having an easy to use metrics pipeline would be very helpful to evaluate impact.

**Comment by LinEricYang** [comment]
@vishh The static policy is our first step to enable CAT in k8s. It is trying to benefit the most of user case without any performance regression. And User doesn't have to do careful benchmark before having cache control functionality. Absolutely we will create a dynamical policy as next step to make it smarter, like a feedback algorithm to adjust the CLOS size based on the monitoring metrics.

You can find the whitepaper about how CAT benefit NFV case [here.](https://builders.intel.com/docs/networkbuilders/deterministic_network_functions_virtualization_with_Intel_Resource_Director_Technology.pdf)

File: contributors/design-proposals/node/cache-management.md (hunk 4 above)
Comment (LinEricYang) [reply contributors/design-proposals/node/cache-management.md]: True. We are working on another proposal to integrate cache monitoring technology with cadvisor, which will allow user to monitor the cache usage of the specific container.

File: contributors/design-proposals/node/cache-management.md (hunk 3 above)
Comment (LinEricYang) [reply contributors/design-proposals/node/cache-management.md]: Thanks for pointing out this, @vishh. Absolutely two logical thread will share the L1 and L2 cache in one cpu core. Will re-work this part.

File: contributors/design-proposals/node/cache-management.md (hunk 2 above)
Comment (dakshinai) [reply contributors/design-proposals/node/cache-management.md]: A small clarification here, though cache allocation happens alongside CPU pinning done by CPU manager, we do not exactly rely on how CPU manager assigns cores to a workload. We would map cache ways to container process id's. This is a feature through resctrl fs on kernel version >= 4.10

**Comment by linericyang** [comment]
We did the benchmark to show the performance benefit of this proposal. The result shows it can reduce half of cache latency when the high-priority workload co-locates with the noisy neighbor. More result detail can be found [here](https://docs.google.com/presentation/d/1t2wB9i07CobU41MPT5-mvyl-uS3v8QykMSjGT8rDcIU/)

@ConnorDoyle @LuyaoZhong

**Pull request closed without merging**